### ISSUE TITLE
Add versioning / provenance tool/tab for graphs

### ISSUE THREAD

**Issue by jlaw9** [Murali-group/GraphSpace#319]
We should track every update made to a graph
- initial post
- update to graph
- update to metadata
- layout added(?)

**Comment by jahandaniyal** [comment]
Hi @jlaw9  Could you please elaborate the requirements of provenance tool/tab you mentioned. 
Will the versioning tab serve only as a log of changes made or should it allow user to go back to the state of the Graph?

**Comment by jlaw9** [comment]
Hi @jahandaniyal Just saw you were working on #336. This is another step in the "git for graphs" project. We are planning on versioning graphs and allowing users to go back to previous states of the graph as well as features like migrate a layout from a previous version.

**Comment by adbharadwaj** [comment]
@jahandaniyal I will answer your query in two parts. 

**First**, the overall goal of this issue is to implement a simple version of "Git for Graphs", which will allow users to track changes made to a graph, to describe these changes, as well as to easily undo these changes. This feature will help users to document the process and workflow that produced the given graph. Here is a list of features that the versioning system must support:

- **Versioning** - This is the main goal of the project. Users should be able to create and save different versions of the same graph. They should provide a "commit" message that describes the changes in the given graph with respect to previous graphs. Each version should have a unique link.

- **Backtracking** - Users should be able to go back to previous versions at any point of time using graphical interface and RESTful interface.

- **Layout compatibility** - Since there may be different versions of the same graph, a layout that was created for one version may not be compatible with a later version. In order to support layout compatibility, we propose to add a feature which will allow users to:
    - Transfer layouts from one version to another. This action is similar to creating a pull request in Github.
    - Layout compatibility check. If the layout is not compatibility, the tool should report the conflict. This check is similar to Github’s feature to check if a pull request is ready to merge or not.


**Second**, We dont just want to log changes, we want the users to be able to go back to previous versions just like Git.

**Comment by jahandaniyal** [comment]
@jlaw9 Indeed, I was working on #336. I am planning to apply for `Git for graphs` project for GSOC '18 and this issue seems to be related, hence my interest.
@adbharadwaj Thank you. Can I work on this issue? Maybe make a POC like I did for #336?.

**Comment by tmmurali** [comment]
@jahandaniyal a proof of concept implementation will be very welcome. Thanks for the initiative.
